content/doc/book/pipeline/jenkinsfile.adoc

```
@@ -889,6 +853,177 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
+This occurs when a user-controller variable (generally an environment variable, usually a parameter passed to the build) that contains special characters (e.g. `' " / \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps using Groovy interpolation.
+For a simple example:
+
+```
+pipeline {
+  agent any
+  parameters {
+    password(name: 'EXAMPLE_SECRET', defaultValue: 'sec>ret', description: 'Enter a password')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        sh "echo $EXAMPLE_SECRET"
+      }
+    }
+  }
+}
+```
+In this example, the argument to the `sh` step is evaluated by Groovy and `SECRET_VALUE` is interpolated directly into the argument as if `sh('echo sec>ret')` has been written in the Pipeline.
+When this is processed on the agent, rather than echoing the value `sec>ret`, it will create a file called `ret` containing the value `sec`.
+Any user able to control a variable interpolated by such a step would be able to make the `sh` step run arbitrary code on the agent.
+To avoid this problem, make sure arguments to steps such as `sh` or `bat` that reference parameters or other user-controller environment variables use single quotes to avoid Groovy interpolation.
+
+```
+pipeline {
+  agent any
+  parameters {
+    password(name: 'EXAMPLE_SECRET', defaultValue: 'sec>ret', description: 'Enter a password')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        sh 'echo $EXAMPLE_SECRET'
+      }
+    }
+  }
+}
+```
+Credential mangling is another issue that can occur when credentials that contain special characters are passed to a step using Groovy interpolation.
+When the credential value is mangled, it is no longer valid and will no longer be masked in the console log.
+
+.NOTE: When using password parameters, the Masked Passwords plugin should be installed to mask sensitive parameters from the console
+```
+pipeline {
+  agent any
+  parameters {
+    password(name: 'EXAMPLE_SECRET', defaultValue: 'sec%ret', description: 'Enter a password')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        bat "echo $EXAMPLE_SECRET"
+      }
+    }
+  }
+}
+```
+Here, the `bat` step receives `echo sec%ret` and the Windows batch shell will simply drop the `%` and print out the value `secret`.
+Because there is a single character difference, the value `secret` will not be masked.
+Though the value is not the same as the actual password, this is still a significant exposure of sensitive information.
+Again, single-quotes avoids this issue.
+
+.NOTE: When using password parameters, the Masked Passwords plugin should be installed to mask sensitive parameters from the console
+```
+pipeline {
+  agent any
+  parameters {
+    password(name: 'EXAMPLE_SECRET', defaultValue: 'sec%ret', description: 'Enter a password')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        bat 'echo $EXAMPLE_SECRET'
```

Windows uses `%` instead of `$` (also, should there be quotes around `%foo%` in case it has spaces? I don't think it matters for `echo`, but maybe it does in other cases? I can never remember how Windows handles this stuff):

```suggestion
        bat 'echo %EXAMPLE_SECRET%'
```

---

fixed in 3481e9984ce629aa54dab6a8d64dac087f3fc4b3